File: code/csharp/ch09/luavm/state/LuaStack.cs
using System;
using System.Linq;
using System.Reflection;
using System.Threading;
using luavm.api;

namespace luavm.state
{
    public class LuaStack
    {
        public object[] slots;
        internal int top;
        internal LuaStack prev;
        internal Closure closure;
        internal object[] varargs;
        internal int pc;
        internal LuaState state;

        internal static LuaStack newLuaStack(int size, LuaState state)
        {
            return new LuaStack
            {
                slots = new object[size],
                top = 0,
                state = state
            };
        }

        internal void check(int n)
        {
            var free = slots.Length - top;
            var slotList = slots.ToList();
            for (var i = free; i < n; i++)
            {
                slotList.Add(null);
            }

            slots = slotList.ToArray();
        }

        internal void push(object val)
        {
            if (top == slots.Length)
            {
                throw new Exception("stack overflow!");
            }

            slots[top] = val;

            top++;
        }

        internal int absIndex(int idx)
        {
            if (idx >= 0 || idx < Consts.LUA_REGISTRYINDEX)
            {
                return idx;
            }

            return idx + top + 1;
        }

        internal bool isValid(int idx)
        {
            if (idx == Consts.LUA_REGISTRYINDEX)
            {
                return true;
            }

            var absIdx = absIndex(idx);
            return absIdx > 0 && absIdx <= top;
        }

        internal object get(int idx)
        {
            if (idx == Consts.LUA_REGISTRYINDEX)
            {
                return state.registry;
            }

            var absIdx = absIndex(idx);
            if (absIdx > 0 && absIdx <= top)
            {
                return slots[absIdx - 1];
            }

            return null;
        }

        internal object pop()
        {
            if (top < 1)
            {
                throw new Exception("stack overflow!");
            }

            top--;
            var val = slots[top];
            slots[top] = null;
            return val;
        }

        internal void pushN(object[] vals, int n)
        {
            var nVals = vals.Length;
            if (n < 0)
            {
                n = nVals;
            }

            for (var i = 0; i < n; i++)
            {
                if (i < nVals)
                {
                    push(vals[i]);
                }
                else
                {
                    push(null);
                }
            }
        }

        internal object[] popN(int n)
        {
            var vals = new object[n];
            for (var i = n - 1; i >= 0; i--)
            {
                vals[i] = pop();
            }

            return vals;
        }


        internal void set(int idx, object val)
        {
            if (idx == Consts.LUA_REGISTRYINDEX)
            {
                state.registry = (LuaTable) val;
                return;
            }

            var absIdx = absIndex(idx);
            if (absIdx <= 0 || absIdx > top)
            {
                throw new Exception("invalid index!");
            }

            slots[absIdx - 1] = val;
        }

        internal void reverse(int from, int to)
        {
            if (to > from)
            {
                Array.Reverse(slots, from, to - from + 1);
            }
            else if (to < from)
            {
                Array.Reverse(slots, to, from - to + 1);
            }
        }
    }
}

File: code/python/ch09/src/lua_stack.py
from consts import Consts


class LuaStack:
    MAX_STACK_SIZE = 1000

    def __init__(self, lua_state):
        self.slots = []
        self.closure = None
        self.varargs = None
        self.pc = 0
        self.caller = None
        self.lua_state = lua_state

    def top(self):
        return len(self.slots)

    def check(self, n):
        return len(self.slots) + n <= LuaStack.MAX_STACK_SIZE

    def push(self, val):
        if len(self.slots) > LuaStack.MAX_STACK_SIZE:
            raise Exception('Stack Over Flow')
        self.slots.append(val)

    def pop(self):
        ret = self.slots[-1]
        self.slots.pop()
        return ret

    def abs_index(self, idx):
        if idx <= Consts.LUA_REGISTRYINDEX:
            return idx

        if idx >= 0:
            return idx
        return idx + len(self.slots) + 1

    def is_valid(self, idx):
        if idx == Consts.LUA_REGISTRYINDEX:
            return True

        idx = self.abs_index(idx)
        return (idx > 0) and (idx <= len(self.slots))

    def get(self, idx):
        if idx == Consts.LUA_REGISTRYINDEX:
            return self.lua_state.registry

        if not self.is_valid(idx):
            return None
        return self.slots[self.abs_index(idx)-1]

    def set(self, idx, val):
        if idx == Consts.LUA_REGISTRYINDEX:
            self.lua_state.registry = val
            return

        if not self.is_valid(idx):
            raise Exception('Invalid Index')
        self.slots[self.abs_index(idx)-1] = val

    def reverse(self, begin, end):
        while begin < end:
            self.slots[begin], self.slots[end] = self.slots[end], self.slots[begin]
            begin += 1
            end -= 1

    def pushn(self, vals, n):
        nvals = len(vals)
        if n < 0:
            n = nvals

        for i in range(n):
            self.push(vals[i] if i < nvals else None)

    def popn(self, n):
        vals = []
        for i in range(n):
            vals.append(self.pop())
        vals.reverse()
        return vals
